File: (2)_Pi_Generation/cpp/task.cpp
#include <iostream>
#include <fstream>
#include <cstdlib>
#include <ctime>
#include <chrono>

using namespace std;
using namespace std::chrono;

int main(int argc, char *argv[])
{
    if (argc != 2)
    {
        cout << "Usage: " << argv[0] << " <value_of_l>" << endl;
        return 1;
    }

    int l = atoi(argv[1]);

    double pi, x, y;
    int idumm = -1;
    int lk = 0;

    srand(time(0)); // Seed for random number generation

    // Measure the start time
    auto start = high_resolution_clock::now();

    // Check if the file exists
    ifstream fileCheck("data/pi_values.dat");
    bool fileExists = fileCheck.good();
    fileCheck.close();

    ofstream outFile("data/pi_values.dat", fileExists ? ios::trunc : ios::app); // Open the file with truncation if it exists

    if (!outFile.is_open())
    {
        cout << "Unable to open the file for saving pi." << endl;
        return 1; // Exit the program if the file can't be opened
    }

    for (int i = 1; i <= l; ++i)
    {
        x = rand() / (double)RAND_MAX;
        y = rand() / (double)RAND_MAX;

        if (x * x + y * y <= 1.0)
        {
            lk++;
        }

        pi = lk * 4.0 / i; // Calculate pi at each iteration

        outFile << pi << endl; // Save pi to the file
    }

    outFile.close(); // Close the file after the loop

    // Measure the end time
    auto stop = high_resolution_clock::now();
    auto duration = duration_cast<milliseconds>(stop - start);

    cout << "The values of pi have been written to 'pi_values.dat'." << endl;
    cout << "Execution time: " << duration.count() << " milliseconds." << endl;

    return 0;
}

File: piGeneration/py/task.py
import random
import time
from datetime import datetime
import sys

if len(sys.argv) != 2:
    print("Usage: python task.py <l>")
    sys.exit(1)

l = int(sys.argv[1])
lk = 0

# Rest of your Python script remains unchanged


# Seed for random number generation
random.seed(time.time())

# Measure the start time
start_time = time.time()

file_path = "data/pi_values.dat"

# Check if the file exists
file_exists = False
try:
    with open(file_path, 'r') as file_check:
        file_exists = True
except FileNotFoundError:
    pass

# Open the file with truncation if it exists
with open(file_path, 'w' if file_exists else 'a') as out_file:
    for i in range(1, l + 1):
        x = random.random()
        y = random.random()

        if x**2 + y**2 <= 1.0:
            print(lk)
            lk += 1

        pi = lk * 4.0 / i  # Calculate pi at each iteration
        out_file.write(f"{pi}\n")  # Save pi to the file

# Measure the end time
stop_time = time.time()
# Convert seconds to milliseconds
elapsed_time = (stop_time - start_time) * 1000

print("The values of pi have been written to 'pi_values.dat'.")
print(f"Execution time: {elapsed_time:.2f} milliseconds.")
